### PR TITLE
Optimize index_add for MetaX and Moore Threads

### PR DESCRIPTION
**Pull request by YUANXIGUREN**
### PR Category

Operator, OP Test, Benchmark

### Type of Change

Bug Fix, Performance Optimization

### Description

Optimize `index_add` and `index_add_` for MetaX C550 and Moore Threads S5000.

Both backends add a contiguous-suffix scatter-add path that views dense inputs
as `[prefix, indexed dimension, suffix]` and directly accumulates suffix tiles,
avoiding the generic rank/stride decomposition and intermediate layout work.
Unsupported dtypes, non-contiguous layouts, aliases, and suffix-size-one cases
continue through the existing common/vendor implementation.

The MetaX implementation selects a 1-D flat kernel for `dim == 0` or
`suffix_size < 80`, and a 2-D tiled kernel otherwise. The crossover was
validated across FP32, FP16, and BF16. MetaX also uses FP32 scratch accumulation
when the Triton version does not support BF16 atomic add.

The Moore Threads implementation preserves the existing generic kernel and
adds the contiguous-suffix path for FP32 and FP16. The change also adds
version-aware index-bounds caching, correct lazy-negative index materialization,
and the related `copy` handling required by pointer-based Triton kernels.

Six general contiguous-suffix cases, contention coverage, randomized BF16
correctness, invalid-index checks, and fallback/layout tests are added to the
public test and benchmark coverage. Dispatch does not contain GraphCast names
or exact production shapes.

### Issue

N/A

### Progress

- [ ] Change is properly reviewed.
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Source Identity

| Role | Commit |
|---|---|
| Official / rebased base | `1d40087bd9ea89602b6a89017deef0c2eca4a705` |
| Candidate | `110eefe8c6c4f13804c2cdd28e621e3bdcb59467` |

Both platforms loaded isolated Official and Candidate source snapshots and
validated the runtime operator paths and SHA256 values before accepting results.

### Correctness

| Platform | Result |
|---|---|
| MetaX C550 Official | 103 passed, 4 skipped across `index_add`, `copy`, and `resolve_neg` |
| MetaX C550 Candidate | 213 passed, 4 skipped, including 115 `index_add` and 45 bounds-cache tests |
| Moore Threads S5000 Official | 46 passed, 10 deselected for the supported `index_add` dtype set |
| Moore Threads S5000 Candidate | 82 passed, 5 skipped, 28 deselected; bounds cache 45 passed; focused copy/resolve/lazy-negative 22 passed |

### Environment

| Platform | Python | Torch stack | Triton | Device |
|---|---|---|---|---|
| MetaX C550 | 3.12.11 | torch 2.8.0+metax3.3.0.2 | 3.1.0 | MetaX C550 |
| Moore Threads S5000 | 3.10.12 | torch 2.5.0 | 3.1.0 | MTT S5000 |

### Performance

Speedup is Official FlagGems latency divided by this PR latency. Official and
Candidate use the same benchmark driver, cases, dtypes, warmup, and repetition
settings. The full public set includes both fast-path and fallback control rows.

| Platform | Public scope | All-case geomean | Fast-path geomean | Fast-path minimum |
|---|---|---:|---:|---:|
| MetaX C550 | 2 APIs × 3 dtypes × 13 cases = 78 rows | 3.8978x | 20.2927x (36 rows) | 10.7938x |
| Moore Threads S5000 | 2 APIs × 2 dtypes × 13 cases = 52 rows | 2.393x | 6.710x (24 rows) | 2.251x |

All 36 C550 and all 24 S5000 fast-path rows improve over Official. S5000
formal performance covers FP32 and FP16; BF16 is not claimed for its new path.

The C550 fallback-control subset has a raw 0.9477x geomean. Its single 0.6798x
outlier did not reproduce in eight independent alternating A/B pairs: paired
speedup median was 0.9839x with a 0.9332x–1.0600x range. The original row is
retained rather than replaced. The S5000 fallback subset has a 0.989x geomean
and a 0.887x worst small-case result; these rows do not enter the new kernel.

| Platform | Torch | Official | Candidate | Official/Candidate |
|---|---:|---:|---:|---:|
| MetaX C550 | 56.0996 ms | 169.4042 ms | 39.1055 ms | 4.3320x |
| Moore Threads S5000 | 48.3587 ms | 2406.250 ms | 555.148 ms | 4.334x |

### CODE REVIEW

File: tests/test_copy.py [hunk 1]
```diff
@@ -80,6 +83,63 @@ def test_copy_inplace_broadcast():
...
+    src = torch._neg_view(src_base) if src_negative else src_base
+    dst = torch._neg_view(dst_base) if dst_negative else dst_base
+    expected = torch.tensor(
+        [-1, 2, -3] if src_negative else [1, -2, 3],
+        dtype=dtype,
+        device=flag_gems.device,
+    )
+
+    assert _can_use_triton(dst, src)
+    with flag_gems.use_gems(include=["copy_"]):
```
Comment (103yiran) [inline tests/test_copy.py]: Please don't use `flag_gems.use_gems`, call FlagGems directly. Reference: https://github.com/flagos-ai/FlagGems/blob/master/docs/content/zh-cn/contribution/overview.md#7-算子单元测试-operator-unit-tests
